enhancements/results-forwarding.md

```diff
@@ -0,0 +1,379 @@
+metadata:
+  name: default
+  namespace: openshift-compliance
+generateResultCRDs: false # no CRD-creator container will be provisioned
+storeEvidenceInPVC: false # no PVC store container will be provisioned
+evidenceStorage: null # No storage options are needed
+resultForwarding:
+  provider: grpc
+  grpc:
+    version: v1  # we'll start at v1
+    endpoint: <URL to GRPC service (could be in-cluster or outside of it)>
+    tls:
+      dynamic: 
+      caSecretName: <ref to CA cert/key pair for TLS> . # This is mandatory
+      serverSecretName: <ref to server cert/key pair secret for mTLS>
+      clientSecretName: <ref to client cert/key pair secret for mTLS> # Mandatory if mtls authentication is used
+    authentication:
+      type: <mtls or token>
+      token:  # May be nil if mtls is chosen
+        tokenSecretName: <ref to secret containing token>
+    extraMetadata:  # defined as a map[string]string
+      cluster_name: foo
+      cluster_type: bar
+      randomKey: random value
+...
+```
+
+Note that `generateResultCRDs` is set to `false`, as well as `storeEvidenceInPVC`
+
+```yaml
+evidenceForwarding:
+  provider: s3
+  s3:
+    bucketName: <name of the s3 bucket>
+    objectPrefix:
+    accessKeyID: 
+```
+
+Potential evidence forwarding providers could be implemented with a similar scheme. e.g.
+a potential `http` provider could look as follows:
+
+```yaml
+evidenceForwarding:
+  provider: http
+  http:
+    version: v1
+    endpoint: <URL of the receiving endpoint>
+    tls:
+      caSecretName: <ref to CA cert/key pair for TLS>
+      serverSecretName: <ref to server cert/key pair secret for mTLS>
+      clientSecretName: <ref to client cert/key pair secret for mTLS>
+```
+
+Note that the current implementation proposal will include the `s3` and `pvc`
+providers only.
+
+### Implementation Details/Notes/Constraints
+
+A First step in implementing the forwarding proposal would be to work with the existing
+`v1alpha1` `ScanSetting` object. All changes would need to be done in how the Compliance
+Operator effectuates scans, and thus we'd ensure that backwards compatibility is
+taken into account. A new `crd-creator` pod would be introduced to receive the results
+from the `aggregator` and ensure that the known `ComplianceCheckResult` and
+`ComplianceRemediation` objects are created.
+
+A subsequent step would be introduce the `v1alpha2` `ScanSetting` object, which would
+contain all the tunables needed for the forwarding architecture.
+
+A webhook would be introduced to ensure compatibility between the `v1alpha1` objects
+and the new `v1alpha2` objects.
+
+
+### Risks and Mitigations
+
+`v1alpha1` objects would be easily translated to `v1alpha2` objects, but not the
+other way around. This needs to be thoroughly documented and communicated.
+
+## Design Details
+
+### Open Questions
+
+1. If the configured provider for either the result forwarder or the evidence
+   persistor would be unavailable, do we error the scan entirely?
```

Comment:
I can't see another way. On one hand, you still have the ConfigMaps so you can get some sense of the results, but there's not much else you can do is there? This is essentially the same as if resultserver doesn't come up now.

Reply:
I was wondering similarly, if you've got forwarding enabled and no CRD/PV, what's the workflow like for checking results. Can you still get a brief look at the results without having to run over to the result server?

Reply:
if you're doing forwarding, it's the responsibility of the receiver to be able to display the results. That would be another system entirely. I don't think this needs to be addressed here. One is still able to view results locally if you don't have a custom configuration for GRPC/s3 and you simply use the defaults (with CRD and PVC implementations)